File: backend/app/services/hud_service.py
"""
HUD Compliance Service
Business logic for HUD PRAC compliance operations
"""
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, and_, or_, func
from typing import List, Optional, Dict, Any
from uuid import UUID
from datetime import date, datetime, timedelta
from decimal import Decimal
import logging

from app.models.hud import (
    TenantIncomeCertification, HouseholdMember, IncomeSource,
    UtilityAllowance, REACInspection,
    CertificationType, CertificationStatus, RelationshipType,
    IncomeType, VerificationType, InspectionType, InspectionStatus
)

logger = logging.getLogger(__name__)


class HUDService:
    """Service class for HUD compliance operations"""

    # TENANT INCOME CERTIFICATIONS
    @staticmethod
    async def create_income_certification(db: AsyncSession, org_id: UUID, data: Dict[str, Any]) -> TenantIncomeCertification:
        """Create a new tenant income certification"""
        certification = TenantIncomeCertification(
            org_id=org_id,
            tenant_id=data["tenant_id"],
            property_id=data["property_id"],
            unit_id=data.get("unit_id"),
            certification_date=data["certification_date"],
            effective_date=data["effective_date"],
            cert_type=data["cert_type"],
            household_size=data["household_size"],
            annual_income=Decimal(str(data["annual_income"])),
            adjusted_income=Decimal(str(data["adjusted_income"])),
            tenant_rent_portion=Decimal(str(data["tenant_rent_portion"])),
            utility_allowance=Decimal(str(data["utility_allowance"])),
            subsidy_amount=Decimal(str(data["subsidy_amount"])),
            certification_status=data.get("certification_status", "pending"),
            created_by=data["created_by"],
        )
        db.add(certification)
        await db.commit()
        await db.refresh(certification)
        return certification

    @staticmethod
    async def get_certifications(
        db: AsyncSession,
        org_id: UUID,
        property_id: Optional[UUID] = None,
        tenant_id: Optional[UUID] = None,
        status: Optional[str] = None,
        cert_type: Optional[str] = None,
    ) -> List[TenantIncomeCertification]:
        """Get tenant income certifications with optional filters"""
        query = select(TenantIncomeCertification).where(
            and_(
                TenantIncomeCertification.org_id == org_id,
                TenantIncomeCertification.deleted_at.is_(None),
            )
        )
        if property_id:
            query = query.where(TenantIncomeCertification.property_id == property_id)
        if tenant_id:
            query = query.where(TenantIncomeCertification.tenant_id == tenant_id)
        if status:
            query = query.where(TenantIncomeCertification.certification_status == status)
        if cert_type:
            query = query.where(TenantIncomeCertification.cert_type == cert_type)
        query = query.order_by(TenantIncomeCertification.effective_date.desc())
        result = await db.execute(query)
        return result.scalars().all()

    @staticmethod
    async def get_certification(db: AsyncSession, cert_id: UUID, org_id: UUID) -> Optional[TenantIncomeCertification]:
        """Get a single income certification by ID"""
        query = select(TenantIncomeCertification).where(
            and_(
                TenantIncomeCertification.id == cert_id,
                TenantIncomeCertification.org_id == org_id,
                TenantIncomeCertification.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        return result.scalar_one_or_none()

    @staticmethod
    async def update_certification(db: AsyncSession, cert_id: UUID, org_id: UUID, data: Dict[str, Any]) -> Optional[TenantIncomeCertification]:
        """Update an existing income certification"""
        certification = await db.execute(
            select(TenantIncomeCertification).where(
                and_(
                    TenantIncomeCertification.id == cert_id,
                    TenantIncomeCertification.org_id == org_id,
                    TenantIncomeCertification.deleted_at.is_(None),
                )
            )
        )
        certification = certification.scalar_one_or_none()
        if not certification:
            return None
        
        for key, value in data.items():
            if hasattr(certification, key):
                if key in ["annual_income", "adjusted_income", "tenant_rent_portion", "utility_allowance", "subsidy_amount"]:
                    setattr(certification, key, Decimal(str(value)))
                else:
                    setattr(certification, key, value)
        
        await db.commit()
        await db.refresh(certification)
        return certification

    @staticmethod
    async def submit_hud_50059(db: AsyncSession, cert_id: UUID, org_id: UUID) -> Optional[TenantIncomeCertification]:
        """Submit HUD 50059 form for certification"""
        certification = await HUDService.get_certification(db, cert_id, org_id)
        if not certification:
            return None
        
        certification.hud_50059_submitted = True
        certification.hud_50059_submission_date = datetime.utcnow()
        certification.certification_status = "approved"
        
        await db.commit()
        await db.refresh(certification)
        return certification

    @staticmethod
    async def get_expiring_certifications(db: AsyncSession, org_id: UUID, days: int = 30) -> List[TenantIncomeCertification]:
        """Get certifications expiring within specified days"""
        cutoff_date = date.today() + timedelta(days=days)
        
        query = select(TenantIncomeCertification).where(
            and_(
                TenantIncomeCertification.org_id == org_id,
                TenantIncomeCertification.effective_date <= cutoff_date,
                TenantIncomeCertification.certification_status == "approved",
                TenantIncomeCertification.deleted_at.is_(None),
            )
        )
        query = query.order_by(TenantIncomeCertification.effective_date.asc())
        result = await db.execute(query)
        return result.scalars().all()

    # HOUSEHOLD MEMBERS
    @staticmethod
    async def add_household_member(db: AsyncSession, tic_id: UUID, org_id: UUID, data: Dict[str, Any]) -> HouseholdMember:
        """Add a household member to a certification"""
        # Verify the certification belongs to the org
        certification = await HUDService.get_certification(db, tic_id, org_id)
        if not certification:
            raise ValueError("Certification not found")
        
        member = HouseholdMember(
            tic_id=tic_id,
            full_name=data["full_name"],
            ssn_last_4=data.get("ssn_last_4"),
            date_of_birth=data["date_of_birth"],
            relationship_type=data["relationship_type"],
            is_student=data.get("is_student", False),
            is_disabled=data.get("is_disabled", False),
            annual_income=Decimal(str(data.get("annual_income", 0))),
        )
        db.add(member)
        await db.commit()
        await db.refresh(member)
        return member

    @staticmethod
    async def update_household_member(db: AsyncSession, member_id: UUID, org_id: UUID, data: Dict[str, Any]) -> Optional[HouseholdMember]:
        """Update a household member"""
        # Get member with certification verification
        query = select(HouseholdMember).join(TenantIncomeCertification).where(
            and_(
                HouseholdMember.id == member_id,
                TenantIncomeCertification.org_id == org_id,
                HouseholdMember.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        member = result.scalar_one_or_none()
        if not member:
            return None
        
        for key, value in data.items():
            if hasattr(member, key):
                if key == "annual_income":
                    setattr(member, key, Decimal(str(value)))
                else:
                    setattr(member, key, value)
        
        await db.commit()
        await db.refresh(member)
        return member

    @staticmethod
    async def remove_household_member(db: AsyncSession, member_id: UUID, org_id: UUID) -> bool:
        """Soft delete a household member"""
        query = select(HouseholdMember).join(TenantIncomeCertification).where(
            and_(
                HouseholdMember.id == member_id,
                TenantIncomeCertification.org_id == org_id,
                HouseholdMember.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        member = result.scalar_one_or_none()
        if not member:
            return False
        
        member.deleted_at = datetime.utcnow()
        await db.commit()
        return True

    # INCOME SOURCES
    @staticmethod
    async def add_income_source(db: AsyncSession, member_id: UUID, org_id: UUID, data: Dict[str, Any]) -> IncomeSource:
        """Add an income source to a household member"""
        # Verify the member belongs to a certification in the org
        query = select(HouseholdMember).join(TenantIncomeCertification).where(
            and_(
                HouseholdMember.id == member_id,
                TenantIncomeCertification.org_id == org_id,
                HouseholdMember.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        member = result.scalar_one_or_none()
        if not member:
            raise ValueError("Household member not found")
        
        income_source = IncomeSource(
            household_member_id=member_id,
            income_type=data["income_type"],
            employer_name=data.get("employer_name"),
            monthly_amount=Decimal(str(data["monthly_amount"])),
            annual_amount=Decimal(str(data["annual_amount"])),
            verification_type=data["verification_type"],
            verification_date=data["verification_date"],
        )
        db.add(income_source)
        await db.commit()
        await db.refresh(income_source)
        return income_source

    @staticmethod
    async def update_income_source(db: AsyncSession, source_id: UUID, org_id: UUID, data: Dict[str, Any]) -> Optional[IncomeSource]:
        """Update an income source"""
        query = select(IncomeSource).join(HouseholdMember).join(TenantIncomeCertification).where(
            and_(
                IncomeSource.id == source_id,
                TenantIncomeCertification.org_id == org_id,
                IncomeSource.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        income_source = result.scalar_one_or_none()
        if not income_source:
            return None
        
        for key, value in data.items():
            if hasattr(income_source, key):
                if key in ["monthly_amount", "annual_amount"]:
                    setattr(income_source, key, Decimal(str(value)))
                else:
                    setattr(income_source, key, value)
        
        await db.commit()
        await db.refresh(income_source)
        return income_source

    @staticmethod
    async def remove_income_source(db: AsyncSession, source_id: UUID, org_id: UUID) -> bool:
        """Soft delete an income source"""
        query = select(IncomeSource).join(HouseholdMember).join(TenantIncomeCertification).where(
            and_(
                IncomeSource.id == source_id,
                TenantIncomeCertification.org_id == org_id,
                IncomeSource.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        income_source = result.scalar_one_or_none()
        if not income_source:
            return False
        
        income_source.deleted_at = datetime.utcnow()
        await db.commit()
        return True

    # UTILITY ALLOWANCES
    @staticmethod
    async def get_utility_allowances(db: AsyncSession, org_id: UUID, property_id: Optional[UUID] = None) -> List[UtilityAllowance]:
        """Get utility allowances for a property"""
        query = select(UtilityAllowance).where(
            and_(
                UtilityAllowance.org_id == org_id,
                UtilityAllowance.deleted_at.is_(None),
            )
        )
        if property_id:
            query = query.where(UtilityAllowance.property_id == property_id)
        query = query.order_by(UtilityAllowance.effective_date.desc(), UtilityAllowance.bedroom_count.asc())
        result = await db.execute(query)
        return result.scalars().all()

    @staticmethod
    async def create_utility_allowance(db: AsyncSession, org_id: UUID, data: Dict[str, Any]) -> UtilityAllowance:
        """Create a new utility allowance"""
        allowance = UtilityAllowance(
            org_id=org_id,
            property_id=data["property_id"],
            bedroom_count=data["bedroom_count"],
            heating=Decimal(str(data.get("heating", 0))),
            cooking=Decimal(str(data.get("cooking", 0))),
            lighting=Decimal(str(data.get("lighting", 0))),
            water_sewer=Decimal(str(data.get("water_sewer", 0))),
            trash=Decimal(str(data.get("trash", 0))),
            total_allowance=Decimal(str(data["total_allowance"])),
            effective_date=data["effective_date"],
        )
        db.add(allowance)
        await db.commit()
        await db.refresh(allowance)
        return allowance

    @staticmethod
    async def get_current_allowance(db: AsyncSession, org_id: UUID, property_id: UUID, bedroom_count: int) -> Optional[UtilityAllowance]:
        """Get current utility allowance for a property and bedroom count"""
        today = date.today()
        query = select(UtilityAllowance).where(
            and_(
                UtilityAllowance.org_id == org_id,
                UtilityAllowance.property_id == property_id,
                UtilityAllowance.bedroom_count == bedroom_count,
                UtilityAllowance.effective_date <= today,
                UtilityAllowance.deleted_at.is_(None),
            )
        )
        query = query.order_by(UtilityAllowance.effective_date.desc()).limit(1)
        result = await db.execute(query)
        return result.scalar_one_or_none()

    # REAC INSPECTIONS
    @staticmethod
    async def get_inspections(db: AsyncSession, org_id: UUID, property_id: Optional[UUID] = None) -> List[REACInspection]:
        """Get REAC inspections for properties"""
        query = select(REACInspection).join(TenantIncomeCertification.property).where(
            and_(
                TenantIncomeCertification.org_id == org_id,
                REACInspection.deleted_at.is_(None),
            )
        )
        if property_id:
            query = query.where(REACInspection.property_id == property_id)
        query = query.order_by(REACInspection.inspection_date.desc())
        result = await db.execute(query)
        return result.scalars().all()

    @staticmethod
    async def create_inspection(db: AsyncSession, org_id: UUID, data: Dict[str, Any]) -> REACInspection:
        """Create a new REAC inspection record"""
        inspection = REACInspection(
            property_id=data["property_id"],
            inspection_date=data["inspection_date"],
            inspection_type=data["inspection_type"],
            overall_score=data.get("overall_score"),
            inspection_status=data["inspection_status"],
            deficiencies_count=data.get("deficiencies_count", 0),
            critical_deficiencies=data.get("critical_deficiencies", 0),
            report_url=data.get("report_url"),
            next_inspection_date=data.get("next_inspection_date"),
        )
        db.add(inspection)
        await db.commit()
        await db.refresh(inspection)
        return inspection

    @staticmethod
    async def update_inspection(db: AsyncSession, inspection_id: UUID, org_id: UUID, data: Dict[str, Any]) -> Optional[REACInspection]:
        """Update a REAC inspection"""
        # Verify inspection belongs to org through property relationship
        query = select(REACInspection).join(TenantIncomeCertification.property).where(
            and_(
                REACInspection.id == inspection_id,
                TenantIncomeCertification.org_id == org_id,
                REACInspection.deleted_at.is_(None),
            )
        )
        result = await db.execute(query)
        inspection = result.scalar_one_or_none()
        if not inspection:
            return None
        
        for key, value in data.items():
            if hasattr(inspection, key):
                setattr(inspection, key, value)
        
        await db.commit()
        await db.refresh(inspection)
        return inspection

    @staticmethod
    async def get_upcoming_inspections(db: AsyncSession, org_id: UUID, days: int = 60) -> List[REACInspection]:
        """Get upcoming REAC inspections within specified days"""
        cutoff_date = date.today() + timedelta(days=days)
        
        query = select(REACInspection).join(TenantIncomeCertification.property).where(
            and_(
                TenantIncomeCertification.org_id == org_id,
                REACInspection.next_inspection_date <= cutoff_date,
                REACInspection.next_inspection_date >= date.today(),
                REACInspection.deleted_at.is_(None),
            )
        )
        query = query.order_by(REACInspection.next_inspection_date.asc())
        result = await db.execute(query)
        return result.scalars().all()

    # RENT CALCULATIONS
    @staticmethod
    async def calculate_tenant_rent(
        household_income: Decimal,
        household_size: int,
        utility_allowance: Decimal,
        contract_rent: Decimal
    ) -> Dict[str, Any]:
        """Calculate tenant rent portion and subsidy amount"""
        # HUD rent calculation: 30% of adjusted monthly income
        monthly_income = household_income / Decimal("12")
        tenant_rent_portion = monthly_income * Decimal("0.30")
        
        # Subtract utility allowance from tenant portion
        tenant_rent_after_utilities = max(Decimal("0"), tenant_rent_portion - utility_allowance)
        
        # Calculate subsidy amount
        subsidy_amount = max(Decimal("0"), contract_rent - tenant_rent_after_utilities)
        
        return {
            "tenant_rent": str(tenant_rent_after_utilities),
            "subsidy_amount": str(subsidy_amount),
            "total_contract_rent": str(contract_rent),
            "utility_allowance": str(utility_allowance),
            "monthly_income": str(monthly_income),
            "rent_to_income_ratio": str((tenant_rent_after_utilities / monthly_income * 100) if monthly_income > 0 else 0),
        }

File: frontend/src/services/hud.service.ts
/**
 * HUD Compliance Service
 * TypeScript API client for HUD compliance endpoints
 */
import {
  TenantIncomeCertification,
  HouseholdMember,
  IncomeSource,
  UtilityAllowance,
  REACInspection,
  CreateTenantIncomeCertificationRequest,
  UpdateTenantIncomeCertificationRequest,
  CreateHouseholdMemberRequest,
  UpdateHouseholdMemberRequest,
  CreateIncomeSourceRequest,
  UpdateIncomeSourceRequest,
  CreateUtilityAllowanceRequest,
  UpdateUtilityAllowanceRequest,
  CreateREACInspectionRequest,
  UpdateREACInspectionRequest,
  CalculateRentRequest,
  CalculateRentResponse,
  CertificationFilters,
  HouseholdMemberFilters,
  IncomeSourceFilters,
  UtilityAllowanceFilters,
  REACInspectionFilters,
  ExpiringCertificationsRequest,
  UpcomingInspectionsRequest,
  CurrentUtilityAllowanceRequest,
  ApiResponse,
  ApiListResponse
} from '@/types/hud';

class HUDServiceError extends Error {
  constructor(
    message: string,
    public status?: number,
    public response?: any
  ) {
    super(message);
    this.name = 'HUDServiceError';
  }
}

export class HUDService {
  private baseUrl = `${process.env.NEXT_PUBLIC_API_URL}/api/v1`;
  private basePath = '/hud';

  private async getAuthHeaders(): Promise<HeadersInit> {
    const token = localStorage.getItem('access_token');
    if (!token) {
      throw new HUDServiceError('No authentication token found');
    }

    return {
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json',
    };
  }

  private async handleResponse<T>(response: Response): Promise<T> {
    if (!response.ok) {
      let errorMessage = `HTTP ${response.status}: ${response.statusText}`;
      try {
        const errorData = await response.json();
        errorMessage = errorData.detail || errorMessage;
      } catch {
        // Use default error message if JSON parsing fails
      }
      throw new HUDServiceError(errorMessage, response.status);
    }

    const data = await response.json();
    return data.data;
  }

  private buildQueryString(params: Record<string, any>): string {
    const searchParams = new URLSearchParams();
    Object.entries(params).forEach(([key, value]) => {
      if (value !== undefined && value !== null) {
        searchParams.append(key, value.toString());
      }
    });
    return searchParams.toString();
  }

  // Tenant Income Certifications
  async getCertifications(filters?: CertificationFilters): Promise<TenantIncomeCertification[]> {
    const queryString = filters ? `?${this.buildQueryString(filters)}` : '';
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications${queryString}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<TenantIncomeCertification[]>(response);
  }

  async createCertification(data: CreateTenantIncomeCertificationRequest): Promise<TenantIncomeCertification> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<TenantIncomeCertification>(response);
  }

  async getCertification(id: string): Promise<TenantIncomeCertification> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications/${id}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<TenantIncomeCertification>(response);
  }

  async updateCertification(id: string, data: UpdateTenantIncomeCertificationRequest): Promise<TenantIncomeCertification> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications/${id}`, {
      method: 'PUT',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<TenantIncomeCertification>(response);
  }

  async submitHUD50059(id: string): Promise<TenantIncomeCertification> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications/${id}/submit`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<TenantIncomeCertification>(response);
  }

  async getExpiringCertifications(days: number = 30): Promise<TenantIncomeCertification[]> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications/expiring?days=${days}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<TenantIncomeCertification[]>(response);
  }

  // Household Members
  async addHouseholdMember(certId: string, data: CreateHouseholdMemberRequest): Promise<HouseholdMember> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/certifications/${certId}/members`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<HouseholdMember>(response);
  }

  async updateHouseholdMember(id: string, data: UpdateHouseholdMemberRequest): Promise<HouseholdMember> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/members/${id}`, {
      method: 'PUT',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<HouseholdMember>(response);
  }

  async removeHouseholdMember(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/members/${id}`, {
      method: 'DELETE',
      headers: await this.getAuthHeaders(),
    });
    if (!response.ok) {
      throw new HUDServiceError(`Failed to remove household member: ${response.statusText}`, response.status);
    }
  }

  // Income Sources
  async addIncomeSource(memberId: string, data: CreateIncomeSourceRequest): Promise<IncomeSource> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/members/${memberId}/income`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<IncomeSource>(response);
  }

  async updateIncomeSource(id: string, data: UpdateIncomeSourceRequest): Promise<IncomeSource> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/income/${id}`, {
      method: 'PUT',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<IncomeSource>(response);
  }

  async removeIncomeSource(id: string): Promise<void> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/income/${id}`, {
      method: 'DELETE',
      headers: await this.getAuthHeaders(),
    });
    if (!response.ok) {
      throw new HUDServiceError(`Failed to remove income source: ${response.statusText}`, response.status);
    }
  }

  // Utility Allowances
  async getUtilityAllowances(propertyId?: string): Promise<UtilityAllowance[]> {
    const queryString = propertyId ? `?property_id=${propertyId}` : '';
    const response = await fetch(`${this.baseUrl}${this.basePath}/utility-allowances${queryString}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<UtilityAllowance[]>(response);
  }

  async createUtilityAllowance(data: CreateUtilityAllowanceRequest): Promise<UtilityAllowance> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/utility-allowances`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<UtilityAllowance>(response);
  }

  async getCurrentAllowance(propertyId: string, bedroomCount: number): Promise<UtilityAllowance> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/utility-allowances/current?property_id=${propertyId}&bedroom_count=${bedroomCount}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<UtilityAllowance>(response);
  }

  // REAC Inspections
  async getInspections(propertyId?: string): Promise<REACInspection[]> {
    const queryString = propertyId ? `?property_id=${propertyId}` : '';
    const response = await fetch(`${this.baseUrl}${this.basePath}/inspections${queryString}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<REACInspection[]>(response);
  }

  async createInspection(data: CreateREACInspectionRequest): Promise<REACInspection> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/inspections`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<REACInspection>(response);
  }

  async updateInspection(id: string, data: UpdateREACInspectionRequest): Promise<REACInspection> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/inspections/${id}`, {
      method: 'PUT',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<REACInspection>(response);
  }

  async getUpcomingInspections(days: number = 60): Promise<REACInspection[]> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/inspections/upcoming?days=${days}`, {
      method: 'GET',
      headers: await this.getAuthHeaders(),
    });
    return this.handleResponse<REACInspection[]>(response);
  }

  // Rent Calculations
  async calculateRent(data: CalculateRentRequest): Promise<CalculateRentResponse> {
    const response = await fetch(`${this.baseUrl}${this.basePath}/calculate-rent`, {
      method: 'POST',
      headers: await this.getAuthHeaders(),
      body: JSON.stringify(data),
    });
    return this.handleResponse<CalculateRentResponse>(response);
  }

  // Utility methods for common operations
  async getCertificationWithMembers(id: string): Promise<TenantIncomeCertification> {
    const certification = await this.getCertification(id);
    // The API already returns household members with the certification
    return certification;
  }

  async getCompleteCertification(id: string): Promise<TenantIncomeCertification> {
    const certification = await this.getCertification(id);
    // The API returns complete certification with all nested data
    return certification;
  }

  // Batch operations for efficiency
  async createCertificationWithMembers(certificationData: CreateTenantIncomeCertificationRequest, membersData: CreateHouseholdMemberRequest[]): Promise<TenantIncomeCertification> {
    // Create certification first
    const certification = await this.createCertification(certificationData);
    
    // Add all household members
    const members = await Promise.all(
      membersData.map(memberData => 
        this.addHouseholdMember(certification.id, memberData)
      )
    );
    
    // Return certification with members
    return {
      ...certification,
      household_members: members
    } as TenantIncomeCertification;
  }

  async addMemberWithIncomeSources(certId: string, memberData: CreateHouseholdMemberRequest, incomeSourcesData: CreateIncomeSourceRequest[]): Promise<HouseholdMember> {
    // Add household member first
    const member = await this.addHouseholdMember(certId, memberData);
    
    // Add all income sources for this member
    const incomeSources = await Promise.all(
      incomeSourcesData.map(incomeData => 
        this.addIncomeSource(member.id, incomeData)
      )
    );
    
    // Return member with income sources
    return {
      ...member,
      income_sources: incomeSources
    } as HouseholdMember;
  }

  // Dashboard and summary methods
  async getCertificationSummary(): Promise<{
    total: number;
    pending: number;
    approved: number;
    expiring: number;
  }> {
    const [allCertifications, expiringCertifications] = await Promise.all([
      this.getCertifications(),
      this.getExpiringCertifications(30)
    ]);

    const pending = allCertifications.filter(c => c.certification_status === 'pending').length;
    const approved = allCertifications.filter(c => c.certification_status === 'approved').length;

    return {
      total: allCertifications.length,
      pending,
      approved,
      expiring: expiringCertifications.length
    };
  }

  async getInspectionSummary(): Promise<{
    total: number;
    passed: number;
    failed: number;
    conditional: number;
    upcoming: number;
    averageScore: number;
  }> {
    const [allInspections, upcomingInspections] = await Promise.all([
      this.getInspections(),
      this.getUpcomingInspections(60)
    ]);

    const passed = allInspections.filter(i => i.inspection_status === 'passed').length;
    const failed = allInspections.filter(i => i.inspection_status === 'failed').length;
    const conditional = allInspections.filter(i => i.inspection_status === 'conditional').length;
    
    const scoresWithValues = allInspections.filter(i => i.overall_score !== null && i.overall_score !== undefined);
    const averageScore = scoresWithValues.length > 0 
      ? scoresWithValues.reduce((sum, i) => sum + (i.overall_score || 0), 0) / scoresWithValues.length
      : 0;

    return {
      total: allInspections.length,
      passed,
      failed,
      conditional,
      upcoming: upcomingInspections.length,
      averageScore: Math.round(averageScore * 100) / 100
    };
  }

  // Validation helpers
  validateCertificationData(data: CreateTenantIncomeCertificationRequest): string[] {
    const errors: string[] = [];
    
    if (!data.tenant_id) errors.push('Tenant ID is required');
    if (!data.property_id) errors.push('Property ID is required');
    if (!data.certification_date) errors.push('Certification date is required');
    if (!data.effective_date) errors.push('Effective date is required');
    if (!data.cert_type) errors.push('Certification type is required');
    if (!data.household_size || data.household_size < 1) errors.push('Household size must be at least 1');
    if (!data.annual_income || parseFloat(data.annual_income) < 0) errors.push('Annual income must be non-negative');
    if (!data.adjusted_income || parseFloat(data.adjusted_income) < 0) errors.push('Adjusted income must be non-negative');
    if (!data.tenant_rent_portion || parseFloat(data.tenant_rent_portion) < 0) errors.push('Tenant rent portion must be non-negative');
    if (!data.utility_allowance || parseFloat(data.utility_allowance) < 0) errors.push('Utility allowance must be non-negative');
    if (!data.subsidy_amount || parseFloat(data.subsidy_amount) < 0) errors.push('Subsidy amount must be non-negative');
    
    return errors;
  }

  validateHouseholdMemberData(data: CreateHouseholdMemberRequest): string[] {
    const errors: string[] = [];
    
    if (!data.full_name?.trim()) errors.push('Full name is required');
    if (!data.date_of_birth) errors.push('Date of birth is required');
    if (!data.relationship_type) errors.push('Relationship type is required');
    
    // Validate date of birth
    if (data.date_of_birth) {
      const dob = new Date(data.date_of_birth);
      const today = new Date();
      if (dob >= today) errors.push('Date of birth must be in the past');
    }
    
    return errors;
  }

  validateIncomeSourceData(data: CreateIncomeSourceRequest): string[] {
    const errors: string[] = [];
    
    if (!data.income_type) errors.push('Income type is required');
    if (!data.monthly_amount || parseFloat(data.monthly_amount) < 0) errors.push('Monthly amount must be non-negative');
    if (!data.annual_amount || parseFloat(data.annual_amount) < 0) errors.push('Annual amount must be non-negative');
    if (!data.verification_type) errors.push('Verification type is required');
    if (!data.verification_date) errors.push('Verification date is required');
    
    return errors;
  }

  validateUtilityAllowanceData(data: CreateUtilityAllowanceRequest): string[] {
    const errors: string[] = [];
    
    if (!data.property_id) errors.push('Property ID is required');
    if (!data.bedroom_count || data.bedroom_count < 0) errors.push('Bedroom count must be non-negative');
    if (!data.total_allowance || parseFloat(data.total_allowance) < 0) errors.push('Total allowance must be non-negative');
    if (!data.effective_date) errors.push('Effective date is required');
    
    return errors;
  }

  validateInspectionData(data: CreateREACInspectionRequest): string[] {
    const errors: string[] = [];
    
    if (!data.property_id) errors.push('Property ID is required');
    if (!data.inspection_date) errors.push('Inspection date is required');
    if (!data.inspection_type) errors.push('Inspection type is required');
    if (!data.inspection_status) errors.push('Inspection status is required');
    
    if (data.overall_score !== undefined && (data.overall_score < 0 || data.overall_score > 100)) {
      errors.push('Overall score must be between 0 and 100');
    }
    
    return errors;
  }
}

// Export singleton instance
export const hudService = new HUDService();
export default hudService;
